### PR TITLE
hfft and ihfft implementation

### PR DESCRIPTION
**Pull request by aldragan0**
Please see [#1877 ](https://github.com/google/jax/issues/1877) for reference.

### CODE REVIEW

**Comment by aldragan0** [comment]
Some tests for these functions are failing due either the tests using unsupported formats for lax.conj() like np.bool or in the "testFftErrors" because there are no checks made for the axis.

I would really appreciate some guidance or feedback for this.

**Comment by hawkinsp** [comment]
The error I see in the CI is:
```
tests/fft_test.py:201: in <lambda>
    lambda: func(rng([2, 3], dtype=np.float64), axis=[1, 1])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @_wraps(np.fft.hfft)
    def hfft(a, n=None, axis=-1, norm=None):
      conj_a = lax.conj(a)
>     nn = (a.shape[axis] - 1) * 2 if n is None else n
E     TypeError: tuple indices must be integers or slices, not list
```
which is an error in your PR: it appears the test is passing `axis=[1, 1]`, but the implementation of `hfft` does not handle that case.

Note the test case is explicitly testing handling of illegal axis specifications, so the correct thing for your code to do here is to raise an exception if it sees an axis that isn't an integer (which is all `hfft` seems to accept: https://numpy.org/doc/stable/reference/generated/numpy.fft.hfft.html#numpy.fft.hfft ), and modify or extend the tests to expect that error from `hfft`.

**Comment by aldragan0** [comment]
Should I also modify the tests to accept the TypeError thrown by the lax.conj in case invalid input type is given?
[fft_tests.log](https://github.com/google/jax/files/4880147/tests.log)

**Comment by hawkinsp** [comment]
It appears that `numpy.fft.hfft` accepts integer inputs (and casts them to an inexact type). So ideally your implementation of `jax.numpy.fft.hfft` should also; it should not error. There are `promote_to_...` helper functions in `lax_fft.py` that you could use here.

**Comment by hawkinsp** [comment]
You could probably also call `jnp.conj`, which is more liberal in what it accepts than `lax.conj`.

**Comment by aldragan0** [comment]
Thank you, fixed the failing tests now.

File: jax/numpy/fft.py [hunk 1]
```diff
@@ -123,6 +126,22 @@ def irfft(a, n=None, axis=-1, norm=None):
   return _fft_core_1d('irfft', xla_client.FftType.IRFFT, a, s=n, axis=axis,
                       norm=norm)
 
+@_wraps(np.fft.hfft)
+def hfft(a, n=None, axis=-1, norm=None):
+  conj_a = jnp.conj(a)
+  _axis_check_1d('hfft', axis)
+  nn = (a.shape[axis] - 1) * 2 if n is None else n
+  return _fft_core_1d('irfft', xla_client.FftType.IRFFT, conj_a, s=n, axis=axis,
```
Comment (skye) [inline jax/numpy/fft.py:134]:
```suggestion
  return _fft_core_1d('hfft', xla_client.FftType.IRFFT, conj_a, s=n, axis=axis,
```

File: jax/numpy/fft.py [hunk 2]
```diff
@@ -123,6 +126,22 @@ def irfft(a, n=None, axis=-1, norm=None):
...
+                      norm=norm) * nn
+
+@_wraps(np.fft.ihfft)
+def ihfft(a, n=None, axis=-1, norm=None):
+  _axis_check_1d('ihfft', axis)
+  nn = a.shape[axis] if n is None else n
+  output = _fft_core_1d('rfft', xla_client.FftType.RFFT, a, s=n, axis=axis,
```
Comment (skye) [inline jax/numpy/fft.py:141]:
```suggestion
  output = _fft_core_1d('ihfft', xla_client.FftType.RFFT, a, s=n, axis=axis,
```

**Review by skye — CHANGES_REQUESTED** [review]

**Review by skye — APPROVED** [review]

**Comment by skye** [comment]
I'll merge assuming the presubmits pass, thanks!

**Pull request merged**